spark-cassandra-connector/src/main/scala/org/apache/spark/sql/cassandra/DefaultSource.scala

```diff
@@ -0,0 +1,147 @@
+package org.apache.spark.sql.cassandra
+
+import org.apache.spark.sql.{DataFrame, SaveMode, SQLContext}
+import org.apache.spark.sql.SaveMode._
+import org.apache.spark.sql.sources.{CreatableRelationProvider, SchemaRelationProvider, BaseRelation, RelationProvider}
+import org.apache.spark.sql.types.{DataType, StructType}
+
+/**
+ * Cassandra data source extends [[RelationProvider]], [[SchemaRelationProvider]] and [[CreatableRelationProvider]].
+ * It's used internally by Spark SQL to create Relation when create a table which specify the Cassandra data source
+ * e.g.
+ *
+ *      CREATE TEMPORARY TABLE tmpTable
+ *      USING org.apache.spark.sql.cassandra
+ *      OPTIONS (
+ *       c_table "table",
+ *       keyspace "keyspace",
+ *       scan_type "catalyst",
+ *       schema '{"type":"struct","fields":
+ *        [{"name":"a","type":"integer","nullable":true,"metadata":{}},
+ *         {"name":"b","type":"integer","nullable":true,"metadata":{}},
+ *         {"name":"c","type":"integer","nullable":true,"metadata":{}}]}'
+ *      )
+ */
+class DefaultSource extends RelationProvider with SchemaRelationProvider with CreatableRelationProvider {
+
+  /** Creates a new relation for a cassandra table given table, keyspace, and cluster as a parameter.*/
+  override def createRelation(
+               sqlContext: SQLContext,
+               parameters: Map[String, String]): BaseRelation = {
+
+    val option = DataSourceOption(parameters)
+    val schema = option.schemaJsonString.map(DataType.fromJson).map(_.asInstanceOf[StructType])
+    sqlContext.getDataSourceRelation(
+      option.tableName,
+      option.keyspaceName,
+      option.scanType,
+      schema,
+      option.clusterName)
+  }
+
+  /** Creates a new relation for a cassandra table given table, keyspace, cluster and schema as a parameter.*/
+  override def createRelation(
+               sqlContext: SQLContext,
+               parameters: Map[String, String],
+               schema: StructType): BaseRelation = {
+
+    val option = DataSourceOption(parameters)
+    sqlContext.getDataSourceRelation(
+      option.tableName,
+      option.keyspaceName,
+      option.scanType,
+      Option(schema),
+      option.clusterName)
+  }
+
+
+  /** Creates a new relation for a cassandra table given table, keyspace, cluster, schema and data as a parameter.*/
+  override def createRelation(
+               sqlContext: SQLContext,
+               mode: SaveMode,
+               parameters: Map[String, String],
+               data: DataFrame): BaseRelation = {
+
+    val option = DataSourceOption(parameters)
+    val table = sqlContext.getDataSourceRelation(
+      option.tableName,
+      option.keyspaceName,
+      option.scanType,
+      None,
+      option.clusterName)
+
+    val notOverwrite = false
```

Comment:
BTW: Doesn't IDEA underline this and offer an automtic single-click fixup? For me, it does.
```scala
table.insert(data, false)
```